src/emulator/storage/apis/firebase.ts

```diff
@@ -93,13 +93,18 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
   firebaseStorageAPI.get("/b/:bucketId/o/:objectId", async (req, res) => {
     let metadata: StoredFileMetadata;
     let data: Buffer;
+
     try {
       // Both object data and metadata get can use the same handler since they share auth logic.
       ({ metadata, data } = await storageLayer.getObject({
         bucketId: req.params.bucketId,
         decodedObjectId: decodeURIComponent(req.params.objectId),
+        url: `${req.protocol}://${req.hostname}:${req.socket.localPort}`,
         authorization: req.header("authorization"),
         downloadToken: req.query.token?.toString(),
+        urlSignature: req.query["X-Firebase-Signature"] as string,
```

Comment:
We need all three of `X-Firebase-Signature`, `X-Firebase-Date` and `X-Firebase-Expires` to be present in order for us to have a usable signed url. 

It might make sense to add a nested struct:

```
export type SignedUrlParams = {
  urlSignature?: string;
  urlUsableSeconds?: string;
  urlTtlSeconds?: number;
}

export type GetObjectRequest = {
  bucketId: string;
  decodedObjectId: string;
  url?: string;
  authorization?: string;
  downloadToken?: string;
  signedUrl?: SignedUrlParams;
};
```

You could add a helper that validates these parameters that is called inside storage layer's `getObject`:

```
private validateSignedUrl(signedUrl: SignedUrlParams) {...}
```

This helper could do everything you're doing on L257-274. Having a single `SignedUrlParams` struct would make `createUnsignedUrl`'s argument list shorter.

Reply:
also if we make SignedUrlParams optional, i think we should make the parameters in it not optional

Reply:
mhmmm actually nvm ^